src/RoatpCourseManagement/SFA.DAS.RoatpCourseManagement/Application/UkrlpData/GetUkrlpAddressesHandler.cs

```diff
@@ -0,0 +1,148 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using MediatR;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using SFA.DAS.RoatpCourseManagement.Configuration;
+using SFA.DAS.RoatpCourseManagement.InnerApi.Models.Ukrlp;
+using SFA.DAS.RoatpCourseManagement.Services;
+
+namespace SFA.DAS.RoatpCourseManagement.Application.UkrlpData
+{
+    public class GetUkrlpAddressesHandler : IRequestHandler<UkrlpDataCommand, UkprnLookupResponse>
+    {
+        private readonly IUkrlpSoapSerializer _serializer;
+        private readonly HttpClient _httpClient;
+        private readonly ILogger<GetUkrlpAddressesHandler> _logger;
+        private readonly UkrlpApiConfiguration _ukrlpConfiguration;
+        private const int MaximumRecords = 500;
+
+        public GetUkrlpAddressesHandler(IUkrlpSoapSerializer serializer, HttpClient httpClient,
+            ILogger<GetUkrlpAddressesHandler> logger, IOptions<UkrlpApiConfiguration> options)
+        {
+            _serializer = serializer;
+            _httpClient = httpClient;
+            _logger = logger;
+            _ukrlpConfiguration = options.Value;
+        }
+
+        public async Task<UkprnLookupResponse> Handle(UkrlpDataCommand command, CancellationToken cancellationToken)
+        {
+            if (command.ProvidersUpdatedSince != null)
+            {
+                return await ProcessUkrlpSinceLastUpdated(command);
+            }
+
+            var ukprnsToProcess = command.Ukprns;
+            var startValue = 0;
+            var maximumToSet = MaximumRecords;
+            if (maximumToSet > ukprnsToProcess.Count)
+                maximumToSet = ukprnsToProcess.Count;
+
+            var response = new UkprnLookupResponse
+            {
+                Results = new List<ProviderAddress>(),
+                Success = true
+            };
+
+            while (maximumToSet <= ukprnsToProcess.Count && startValue < ukprnsToProcess.Count)
```

Comment:
Linq gives you couple of methods that makes batch processing simpler. 
```
var numbers = Enumerable.Range(1, 50);
const int batchSize = 8;
var fetched = 0;

while (true)
{
	var batch = numbers.Skip(fetched).Take(batchSize);
	if (!batch.Any()) break;
	fetched += batch.Count();
	Console.WriteLine($"Total fetched {fetched} off {numbers.Count()}");
}
```

Reply:
cool, I wasn't aware of this